README.md
# ALToolbox

[![PyPI version](https://badge.fury.io/py/docassemble-ALToolbox.svg)](https://badge.fury.io/py/docassemble-ALToolbox)

This repository is used to host small Python modules, widgets, and JavaScript web components js files that enhance Docassemble interviews. These modules were
built as part of the Suffolk University Law School LIT Lab's [Document Assembly Line project](https://suffolklitlab.org/docassemble-AssemblyLine-documentation/).
They are placed here
rather than in https://github.com/SuffolkLitLab/docassemble-AssemblyLine because we believe these small components can easily be used
by anyone, regardless of whether they use any other code from the Document Assembly Line project.

If you want to add a small function to this project, consider adding it to the existing misc.py to avoid creating too many module files.

## Documentation

Read the [documentation for the functions and components](https://suffolklitlab.org/docassemble-AssemblyLine-documentation/docs/framework/altoolbox) to learn
how to use these components in your own [Docassemble](https://github.com/jhpyle/docassemble) projects.

## Suffolk LIT Lab Document Assembly Line

<img src="https://user-images.githubusercontent.com/7645641/142245862-c2eb02ab-3090-4e97-9653-bb700bf4c54d.png" alt="drawing of people working together on a website UI" width="300" style="align: center;"/>

The Assembly Line Project is a collection of volunteers, students, and institutions who joined together
during the COVID-19 pandemic to help increase access to the court system. Our vision is mobile-friendly,
easy to use **guided** online forms that help empower litigants to access the court remotely.

Our signature project is [CourtFormsOnline.org](https://courtformsonline.org).

We designed a step-by-step, assembly line style process for automating court forms on top of Docassemble
and built several tools along the way that **you** can use in your home jurisdiction.

This package contains **runtime code** and **pre-written questions** to support authoring robust, 
consistent, and attractive Docassemble interviews that help complete court forms.

Read more on our [documentation page](https://suffolklitlab.org/docassemble-AssemblyLine-documentation/).


## Related repositories

* https://github.com/SuffolkLitLab/docassemble-AssemblyLine
* https://github.com/SuffolkLitLab/docassemble-ALWeaver
* https://github.com/SuffolkLitLab/docassemble-ALMassachusetts
* https://github.com/SuffolkLitLab/docassemble-MassAccess
* https://github.com/SuffolkLitLab/docassemble-ALThemeTemplate
* https://github.com/SuffolkLitLab/EfileProxyServer

## Contributors:
* @plocket  
* @nonprofittechy
* @purplesky2016
* @brycestevenwilley


CHANGELOG.md
# CHANGELOG

You can see the changelog of releases for versions 0.4.0 and after at https://github.com/SuffolkLITLab/docassemble-ALToolbox/releases.

Below are the changelogs for [older releases on pypi](https://pypi.org/project/docassemble.ALToolbox/#history).


## Version 0.3.7

### Added

* Added a scrollable display text panel for long acknowledge statements: [#26](https://github.com/SuffolkLITLab/docassemble-ALToolbox/issues/26)

### Fixed

* Patched tabbed_template_html to work with Bootstrap 5 [#37](https://github.com/SuffolkLITLab/docassemble-ALToolbox/pull/37)

### Documentation

* Added a landing page for demos [#34](https://github.com/SuffolkLITLab/docassemble-ALToolbox/issues/34)

[See all changes for 0.3.7](https://github.com/SuffolkLITLab/docassemble-ALToolbox/compare/9e2bee3...476835f)

## Version 0.3.6

### Added

* Added the bootstrap-multiselect input CSS and JS, forked to work with Bootstrap 5 ([see commit 9e2bee](https://github.com/SuffolkLITLab/docassemble-ALToolbox/commit/9e2bee3879c7adf9456675c493810e028cae87ea))

[See all changes for 0.3.6](https://github.com/SuffolkLITLab/docassemble-ALToolbox/compare/91114b6...9e2bee3)

## Version 0.3.5

### Added

* Added `escape_button.yml` based on the version in the 209A ([see commit 91114b6](https://github.com/SuffolkLITLab/docassemble-ALToolbox/commit/91114b67298e01d44d0d5abb90b226f1daf85500))

### Changed

* Modified custom 504 page design and content: [#31](https://github.com/SuffolkLITLab/docassemble-ALToolbox/pull/31)

[See all changes for 0.3.5](https://github.com/SuffolkLITLab/docassemble-ALToolbox/compare/8ea06d5...91114b6)

## Version 0.3.4

### Changed

* Update birthdate selector to bootstrap 5: [#28](https://github.com/SuffolkLITLab/docassemble-ALToolbox/commit/8ea06d51f6ee946a2627990b594c897d98ff227b)

[See all changes for 0.3.4](https://github.com/SuffolkLITLab/docassemble-ALToolbox/compare/5c05305...8ea06d5)

## Version 0.3.3

[No changes.](https://github.com/SuffolkLITLab/docassemble-ALToolbox/compare/c791364...5c05305)

## Version 0.3.2

### Changed

* `collapse_template.yml` is now includeable in interviews, and its demo is at `collapse_template_demo.yml`: ([see commit c791364](https://github.com/SuffolkLITLab/docassemble-ALToolbox/commit/c791364df05866c8654b701f85549fd29f23cfbd))

## Version 0.3.1

### Added

* Added a custom 504 error page: ([see commit e127a1d](https://github.com/SuffolkLITLab/docassemble-ALToolbox/commit/e127a1d49267996250532a80cdb7ee0d9924ac38))

### Fixed

* `collapse_template` works with Bootstrap 5 now: ([see commit 4b1f6a7](https://github.com/SuffolkLITLab/docassemble-ALToolbox/commit/4b1f6a743fd4230613f6d7b0f23097cf993f42a5))

## Version 0.3.0

### Added

* Added `sum_if_defined` function ([see commit 03c3bf5](https://github.com/SuffolkLITLab/docassemble-ALToolbox/commit/03c3bf577cdb5517c4a66844d477979760230710))

## Version 0.2.1

### Fixed

* Render HTML content in templates: ([see commit 35c991e](https://github.com/SuffolkLITLab/docassemble-ALToolbox/commit/35c991e76abbf670ec8b06c3e316e8e5b5f23abd))

## Version 0.2.0

### Added

* Added `tabbed_templates_html` method (only used in EFSP interview now) ([see commit 30db333](https://github.com/SuffolkLITLab/docassemble-ALToolbox/commit/30db333fb11bd56986be641cc4b5967b20f06d4a))

## Version 0.1.0

### Added

* Added `show_decimals` keyword to `thousands` function to show 2 decimals points: ([see commit 744c58e](https://github.com/SuffolkLITLab/docassemble-ALToolbox/commit/744c58edb4c8e9c2b85b54d76914c978038ba8dd))

### Documentation

* Added demo on how to profile a docassemble interview: [#21](https://github.com/SuffolkLITLab/docassemble-ALToolbox/pull/21)

[See all changes for 0.1.0](https://github.com/SuffolkLITLab/docassemble-ALToolbox/compare/be79d30...744c58e)

## Version 0.0.11

[See all changes for 0.0.11](https://github.com/SuffolkLITLab/docassemble-ALToolbox/compare/4178c55...be79d30)

## Version 0.0.10

[See all changes for 0.0.10](https://github.com/SuffolkLITLab/docassemble-ALToolbox/compare/f1cf2e7...4178c55)

## Version 0.0.9.1

[See all changes for 0.0.9.1](https://github.com/SuffolkLITLab/docassemble-ALToolbox/compare/5fa482a...f1cf2e7)

## Version 0.0.9

[No changes](https://github.com/SuffolkLITLab/docassemble-ALToolbox/compare/b578397...5fa482a)

## Version 0.0.8

[See all changes for 0.0.8](https://github.com/SuffolkLITLab/docassemble-ALToolbox/compare/4651661...b578397)

## Version 0.0.7

### Breaking changes

* Changed package name to ALToolbox from AssemblyLineToolbox

[See all changes for 0.0.7](https://github.com/SuffolkLITLab/docassemble-ALToolbox/compare/84782f5...4651661)


docassemble/ALToolbox/misc.py
from typing import Optional, Union

from decimal import Decimal
import docassemble.base.functions
from docassemble.base.util import (
    defined,
    value,
    showifdef,
    space_to_underscore,
    action_button_html,
    Address,
    word,
)
import re

__all__ = [
    "shortenMe",
    "thousands",
    "tel",
    "fa_icon",
    "space",
    "yes_no_unknown",
    "number_to_letter",
    "collapse_template",
    "tabbed_templates_html",
    "review_widget",
    "sum_if_defined",
    "add_records",
    "output_checkbox",
    "nice_county_name",
]


class shortenMe:
    def __init__(self, originalURL):
        self.shortenedURL = docassemble.base.functions.temp_redirect(
            originalURL, 60 * 60 * 24 * 7, False, False
        )


# The following three functions are from Quinten
def thousands(num: Union[float, str, Decimal], show_decimals=False) -> str:
    """
    Return a whole number formatted with thousands separator.
    Optionally, format with 2 decimal points (for a PDF form with the
    currency symbol already present in the form)

    If `show_decimals`, will call `int(num)`, truncating the decimals instead of
    rounding to the closest int.
    """
    try:
        if show_decimals:
            return f"{num:,.2f}"
        else:
            return f"{int(num):,}"
    except:
        return str(num)


def tel(phone_number) -> str:
    """Format a phone number so you can click on it to open in your phone dialer"""
    return '<a href="tel:' + str(phone_number) + '">' + str(phone_number) + "</a>"


def fa_icon(
    icon: str,
    color: Optional[str] = "primary",
    color_css: Optional[str] = None,
    size: Optional[str] = "sm",
    fa_class: str = "fa-solid",
) -> str:
    """Display a fontawesome icon inline.

    Docassemble allows you to display an icon from [fontawesome](https://fontawesome.com),
    but it does not provide control over the size or color of the icon. This function gives
    you more control over the icon that is inserted.

    Args:
      icon: a string representing a fontawesome icon. The icon needs to be in the
        [free library](https://fontawesome.com/search?o=r&m=free).
      color: can be any [Bootstrap color variable](https://getbootstrapc.mo/docs/4.0/utilities/colors).
        For example: `primary`, `secondary`, `warning`
      color_css: allows you to use a CSS code to represent the color, e.g., `blue`, or `#fff` for black
      size: used to control the [fontawesome size](https://fontawesome.com/v6.0/docs/web/style/size)
        (without the `fa-` prefix). Valid values include `2xs`, `xs`, the default of `sm`,
        `md`, `lg`, `xl`, `2x1`, and the python `None`, which defaults to `md`.
      fa_class: let's you specify the fontawesome class, needed for any icon that isn't
        the default class of `fa-solid`, like `fa-brands`, or `fa-regular` and `fa-light`.

    Returns:
      HTML for a font-awesome icon of the specified size and color.
    """
    if not size or size == "md":
        size_str = ""
    else:
        size_str = " fa-" + size
    if not size and not color and not color_css:
        return ":" + icon + ":"  # Default to letting Docassemble handle it
    if color_css:
        return (
            f'<i class="{fa_class} fa-'
            + icon
            + size_str
            + '" style="color:'
            + color_css
            + ';"></i>'
        )
    if color:
        return (
            f'<i class="{fa_class} fa-'
            + icon
            + size_str
            + '" style="color:var(--'
            + color
            + ');"></i>'
        )
    return f'<i class="{fa_class} fa-{icon}{size_str}"></i>'


def space(var_name: str, prefix=" ", suffix="") -> str:
    """If the value as a string is defined, return it prefixed/suffixed. Defaults to prefix
    of a space. Helps build a sentence with less cruft. Equivalent to SPACE function in
    HotDocs."""
    if (
        var_name
        and isinstance(var_name, str)
        and re.search(r"[A-Za-z][A-Za-z0-9\_]*", var_name)
        and defined(var_name)
        and value(var_name)
    ):
        return prefix + showifdef(var_name) + suffix
    else:
        return ""


def yes_no_unknown(
    var_name: str, condition: Optional[bool], unknown="Unknown", placeholder=0
):
    """Return 'unknown' if the value is None rather than False. Helper for PDF filling with
    yesnomaybe fields"""
    if condition:
        return value(var_name)
    elif condition is None:
        return unknown
    else:
        return placeholder


def number_to_letter(n: Optional[int]) -> str:
    """Returns a capital letter representing ordinal position. E.g., 1=A, 2=B, etc. Appends letters
    once you reach 26 in a way compatible with Excel/Google Sheets column naming conventions. 27=AA, 28=AB...
    """
    string = ""
    if n is None:
        n = 0
    while n > 0:
        n, remainder = divmod(n - 1, 26)
        string = chr(65 + remainder) + string
    return string


def collapse_template(
    template,
    classname=None,
    container_classname=None,
    id_tag=None,
    closed_icon="caret-right", 
    open_icon="caret-down",
):
    """
    Insert HTML for a Bootstrap "collapse" div.

    Optionally, you can specify a custom icon to override the defaults:

    The default icons are "right caret" which displays when the text is collapsed (`closed_icon`) and
    "down caret" which displays when the text is open (`open_icon`).
    """
    if not template.subject_as_html(trim=True) and not template.content_as_html():
        return ""
      
    if classname is None:
        classname = " bg-light"
    else:
        classname = " " + classname.strip()
    container_classes_plus = "al_collapse_template"
    if container_classname:
      container_classes_plus += f" { container_classname }"
      
    the_id = re.sub(r"[^A-Za-z0-9]", "", template.instanceName)
    container_id = id_tag or f'{ the_id }_container'
    
    return f"""\
<div id="{ container_id }" class="{ container_classes_plus }">
<a class="collapsed al_toggle" data-bs-toggle="collapse" href="#{ the_id }" role="button" aria-expanded="false" aria-controls="{ the_id }"><span class="toggle-icon pdcaretopen">{ fa_icon(open_icon) }</span><span class="toggle-icon pdcaretclosed">{ fa_icon(closed_icon) }</span>
<span class="subject">{ template.subject_as_html(trim=True) }</span></a>
<div class="collapse" id="{ the_id }">
<div class="card card-body pb-1{ classname }">{ template.content_as_html() }</div>
</div>
</div>\
"""


def tabbed_templates_html(tab_group_name: str, *pargs) -> str:
    """
    Provided a list of templates, create Bootstrap v 4.5 tabs with the `subject` as the tab label.
    """
    if isinstance(tab_group_name, str):
        tab_group_name = space_to_underscore(tab_group_name)
    else:
        tab_group_name = "tabbed-group"
    tabs = f'<ul class="nav nav-tabs" id="{tab_group_name}" role="tablist">\n'
    tab_content = '<div class="tab-content" id="myTabContent">'
    for index, templ in enumerate(pargs):
        tab_id = f"{tab_group_name}-{space_to_underscore(str(templ.subject))}"
        tabs += '<li class="nav-item" role="presentation">\n'
        if index == 0:
            tabs += f'<a class="nav-link active" id="{tab_id}-tab" data-bs-toggle="tab" href="#{tab_id}" role="tab" aria-controls="{tab_id}" aria-selected="true">{templ.subject}</a>\n'
            tab_content += f'<div class="tab-pane fade show active" id="{tab_id}" role="tabpanel" aria-labelledby="{tab_id}-tab">\n'
            tab_content += templ.content_as_html()

            tab_content += "\n</div>\n"
        else:
            tabs += f'<a class="nav-link" id="{tab_id}-tab" data-bs-toggle="tab" href="#{tab_id}" role="tab" aria-controls="{tab_id}" aria-selected="false">{templ.subject}</a>\n'
            tab_content += f'<div class="tab-pane fade" id="{tab_id}" role="tabpanel" aria-labelledby="{tab_id}-tab">\n'
            tab_content += templ.content_as_html()
            tab_content += "\n</div>\n"
        tabs += "</li>"
    tabs += "</ul>"
    tab_content += "</div>"

    return tabs + tab_content


def review_widget(
    *,
    up_action: str,
    down_action: str,
    review_action: Optional[str] = None,
    thumbs_display: str = "Did we help you?",
    review_display: str = "Thank you for your feedback. Let us know what we could do better",
    submit_review_button: str = "Add your review",
    post_review_display: str = "Thank you for your review!",
) -> str:
    """
    A widget that allows people to give a quick review (thumbs up and down, with an optional text
    component) in the middle of an interview without triggering a page reload.

    If `review_action` is provided, once you press either of the thumbs, a text input screen with
    a submit button appears, and once the text review is submitted (or after the thumbs, if no
    `review_action` was provided), a final "thank you" message is displayed.

    Args:
        up_action: the variable name of an event to be executed on the server if the
            thumbs up is pressed
        down_action: the variable name of an event to be executed on the server if the
            thumbs down is pressed
        review_action: the variable name of an event to be execute on the
            server when someone submits their text review
        thumbs_display: text displayed to user describing the thumbs
        review_display: text displayed to user describing the text input
        submit_review_button: text on the button to submit their text review
        post_review_display: text displayed to user after review is submitted
    Returns:
        the HTML string of the widget
    """
    js_thumbs_up = f"javascript:altoolbox_thumbs_up_send('{up_action}', {'true' if review_action else 'false'})"
    js_thumbs_down = f"javascript:altoolbox_thumbs_down_send('{down_action}', {'true' if review_action else 'false'})"
    widget = f"""
<div class="card" style="width: 20rem;">
  <div class="card-body">
    <p class="al-thumbs-widget">{word(thumbs_display)}</p>
    <a href="{js_thumbs_up}" id="al-thumbs-widget-up"
        class="btn btn-md btn-info al-thumbs-widget" aria-label="{word('Thumbs up')}">{fa_icon('thumbs-up', size='md')}</a>
    <a href="{js_thumbs_down}" id="al-thumbs-widget-down"
        class="btn btn-md btn-info al-thumbs-widget" aria-label="{word('Thumbs down')}">{fa_icon('thumbs-down', size='md')}</a>
    """
    if review_action:
        review_area_id = review_action + "_area_id"
        js_review = (
            f"javascript:altoolbox_review_send('{review_action}', '{review_area_id}')"
        )
        widget += f"""
      <p class="al-review-text al-hidden">{word(review_display)}</p>
      <textarea class="datextarea al-review-text al-hidden" id="{review_area_id}"
          alt="{word('Write your review here')}" rows="4"></textarea>
      <br class="al-review-text">
      {action_button_html(js_review, label=word(submit_review_button), size='md', classname='al-review-text al-hidden')} 
        """
    widget += f"""
  <p class="al-after-review al-hidden">{word(post_review_display)}</p>
  </div>
</div>"""
    return widget


def sum_if_defined(*pargs):
    """Lets you add up the value of variables that are not in a list"""
    total = 0
    for source in pargs:
        if defined(source):
            total += value(source)
    return total


def add_records(obj, labels):
    """List demo interviews in the current package to be run from the landing page"""
    index = 0
    for key, val in labels.items():
        obj.appendObject()
        obj[index].name = key
        obj[index].description = val[0]
        obj[index].reference = val[1]
        index += 1
    return obj


def output_checkbox(
    value_to_check: bool, checked_value: str = "[X]", unchecked_value: str = "[  ]"
):
    """Generate a conditional checkbox for docx templates

    Args:
      checked_value: defaults to `[X]` but can be set to any string or even a `DAFile` or `DAStaticFile`
          with the image of a checkbox
      unchecked_value: opposite meaning of `checked_value` and defaults to `[  ]`

    """
    if value_to_check:
        return checked_value
    else:
        return unchecked_value


def nice_county_name(address: Address) -> str:
    """
    If the county name contains the word "County", which Google Address
    Autocomplete does by default, remove it.
    """
    if not hasattr(address, "county"):
        return ""
    if address.county.endswith(" County"):
        return address.county[: -len(" County")]
    else:
        return address.county
